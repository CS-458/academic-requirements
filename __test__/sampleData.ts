export const savedSchedule = [
  {
    year: 0,
    seasons: [
      { season: "Fall", classes: [] },
      { season: "Winter", classes: [] },
      { season: "Spring", classes: [] },
      { season: "Summer", classes: [] }
    ]
  },
  {
    year: 1,
    seasons: [
      { season: "Fall", classes: ["CS-144", "COMST-100", "ENGL-101"] },
      { season: "Winter", classes: [] },
      {
        season: "Spring",
        classes: ["CS-145", "ENGL-102", "MATH-157", "MATH-156"]
      },
      { season: "Summer", classes: [] }
    ]
  },
  {
    year: 2,
    seasons: [
      { season: "Fall", classes: ["CS-244", "CS-248", "MATH-275"] },
      { season: "Winter", classes: [] },
      { season: "Spring", classes: ["CS-254", "CS-324", "MATH-270", "CS-290"] },
      { season: "Summer", classes: [] }
    ]
  },
  {
    year: 3,
    seasons: [
      { season: "Fall", classes: ["CS-343", "ICT-215", "CS-364"] },
      { season: "Winter", classes: [] },
      { season: "Spring", classes: ["CS-358", "CS-404"] },
      { season: "Summer", classes: [] }
    ]
  },
  {
    year: 4,
    seasons: [
      { season: "Fall", classes: ["CS-458", "CS-441"] },
      { season: "Winter", classes: [] },
      { season: "Spring", classes: ["CS-458", "CS-442"] },
      { season: "Summer", classes: [] }
    ]
  }
];
