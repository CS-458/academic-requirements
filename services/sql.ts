import { PromisedDatabase as Database } from "promised-sqlite3";

const academic = {
  db: new Database(),
  opened: false
};

export async function academicDb(): Promise<Database> {
  if (!academic.opened) {
    academic.opened = true;
    await academic.db.open("./database.sqlite");
  }
  return academic.db;
}

export function setAcademicDb(db: Database): void {
  academic.db = db;
}

const user = {
  db: new Database(),
  opened: false
};

export async function userDb(): Promise<Database> {
  if (!user.opened) {
    user.opened = true;
    await user.db.open("./users.sqlite");
  }
  return user.db;
}

export function setUserDb(db: Database): void {
  user.db = db;
  user.opened = true;
}
