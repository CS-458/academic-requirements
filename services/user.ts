import React from "react";
import { ConcentrationType, MajorType } from "../entities/four_year_plan";
import { fetchApi } from "./util";
export interface UserMajor {
  /// Major ID number
  major: MajorType;
  /// Concentration ID number
  concentration: ConcentrationType;
  /// Whether to load a four year plan
  load_four_year_plan: boolean;
  /// A list of completed courses taken
  completed_courses: string[];
}

export function userMajor(): UserMajor | undefined {
  const data = window.localStorage.getItem("user_major");
  if (data !== null) {
    // console.log(data);
    return JSON.parse(data);
  } else {
    return undefined;
  }
}

export function setUserMajor(maj: UserMajor | undefined): void {
  if (maj !== undefined) {
    window.localStorage.setItem("user_major", JSON.stringify(maj));
  } else {
    window.localStorage.removeItem("user_major");
  }
}

export interface UserInfo {
  /// Profile URL
  picture: string;
  /// User ID
  sub: string;
  name: string;
  email: string;
}

export interface User {
  info: UserInfo;
  cred: string;
}

export const UserLogin = React.createContext<undefined | User>(undefined);
export function userToken(): string | undefined {
  return React.useContext(UserLogin)?.cred;
}

// Calls the API to upload the schedule to the Database
export async function uploadSchedule(name: string, schedule: any): Promise<void> {
  const token = userToken();
  if (token === undefined) {
    throw new Error("User is not logged in");
  }
  await fetchApi(`/api/inserts/schedule?name=${name}`, {
    method: "POST",
    body: JSON.stringify(schedule),
    headers: {
      "X-Google-Token": token
    }
  });
}
/*
  Function used to store user into the user table
  May not since we are checking if the user exists
  when uploading the schedule
*/
export async function saveLoggedInUser(): Promise<void> {
  const token = userToken();
  if (token === undefined) {
    throw new Error("User Token Not made");
  }
  await fetchApi("/api/inserts/user", {
    method: "POST",
    headers: {
      "X-Google-Token": token
    }
  });
}
