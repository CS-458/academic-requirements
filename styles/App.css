html {
  print-color-adjust: exact;
  -webkit-print-color-adjust: exact;
}

@media print {
  @page {
    /* size: portrait; */
    size: landscape;
  }
}

.generic {
  height: 100%;
}

.App {
  height: 100%;
  margin-right: 10px;
  margin-left: 10px;
}

.screen {
  height: 100%;
  width: 100%;
  text-align: center;
}

.css-ak1ss8 {
  min-height: 40px;
}

button {
  background-color: #004990;
  color: white;
  text-align: center;
  font-size: 24px;
  padding: 10px;
  border-radius: 15px;
}

body {
  font-family: Verdana, Geneva, Tahoma, sans-serif;
}

.class-dropdown {
  text-align: left;
  border-radius: 10px;
  padding: 5px;
  border-color: black;
  border-width: 2px;
  display: flex;
  flex-direction: column;
  /* background-color: #004990; */
}

.drag-drop {
  height: 100%;
  display: grid;
  grid-template-rows: 100%;
  grid-template-columns: 1fr 58% 25% 10%;
  column-gap: 10px;
}

.schedule {
  height: 100%;
  width: 100%;
  column-gap: 1px;
  row-gap: 1px;
}

.semester:not(:last-child) {
  border-right: 1px lightgray solid;
}

.semester:not(:first-child) {
  border-left: 1px lightgray solid;
}

.page {
  height: 100%;
}

.info-box-completed-course {
  font-size: 15px;
  text-align: center;
  border: 2px solid #004990;
  border-radius: 5px;
  padding: 0.5em;
  margin: 5px;
  color: black;
}

/* Text inside the course dropdown menu*/
.courseDropdowns {
  text-align: center;
  width: 100%;
  float: center;
}

.RequirementText {
  width: 100%;
  padding-bottom: 0;
  display: grid;
  grid-template-columns: 80% 20%;
  text-align: center;
  font-size: 11pt;
  padding-left: 0;
  vertical-align: center;
}
.percentage {
  text-align: right;
  justify-content: right;
  display: flex;
  align-items: center;
}

.requirementName {
  text-align: left;
  display: flex;
  align-items: center;
  padding-top: 1px;
  vertical-align: center;
}
