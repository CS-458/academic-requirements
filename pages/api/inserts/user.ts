/*
    @Description
      This is the API call that pushes the User to the user table
*/

import { NextApiRequest, NextApiResponse } from "next";
import verifyToken from "../../../services/login";
import { userDb } from "../../../services/sql";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method !== "POST") {
    res.status(405).json({ error: "Only POST requests allowed" });
    return;
  }

  // Creates connection to the DB
  const con = await userDb();

  // Headers are lowercased
  const token = req.headers["x-google-token"];
  if (typeof token !== "string") {
    res.status(401).json({ error: "Invalid user not logged in" });
    return;
  }

  // verifyToken handles adding a user if it doesn't already exist
  const user = await verifyToken(token, con);
  if (user === undefined) {
    res.status(401).json({ error: "Invalid user token" });
    return;
  }
  // Returns an error message with invalid parameters
  res.status(200).json({ message: "Successfully registered user" });
}
