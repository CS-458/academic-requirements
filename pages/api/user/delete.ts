/*
    @Description
      This is the API call that pushes the schedule form the user to the Database
*/

import { NextApiRequest, NextApiResponse } from "next";
import verifyToken from "../../../services/login";
import { userDb } from "../../../services/sql";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method !== "POST") {
    res.status(405).json({ error: "Only Post requests allowed" });
    return;
  }

  // Creates connection to the DB
  const con = await userDb();

  // Headers are lowercased
  const token = req.headers["x-google-token"];
  if (typeof token !== "string") {
    res.status(401).json({ error: "Invalid user not logged in" });
    return;
  }

  const user = await verifyToken(token, con);

  if (user === undefined) {
    res.status(401).json({ error: "Invalid user token" });
    return;
  }
  let name = req.query.name;
  if (typeof name !== "string") {
    res.status(400).json({ error: "Invalid name specified" });
    return;
  }
  name = decodeURIComponent(name);

  console.log("Deleting ", name);
  const result = await con.run(
    "DELETE FROM schedule WHERE userID = ? and name = ?",
    [user, name]
  );
  console.log(result);
  if (result.changes === 1) {
    // Returns a success message
    res.status(200).json({});
  } else {
    res.status(404).json({ error: "schedule not found" });
  }
}
