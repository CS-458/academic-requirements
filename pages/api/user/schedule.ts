/*
    @Description
      This is the API call that pushes the schedule form the user to the Database
*/

import { NextApiRequest, NextApiResponse } from "next";
import verifyToken from "../../../services/login";
import sql from "../../../services/sql";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method !== "GET") {
    res.status(405).json({ error: "Only Get requests allowed" });
    return;
  }

  // Creates connection to the DB
  const con = await sql();

  // Headers are lowercased
  const token = req.headers["x-google-token"];
  if (typeof token !== "string") {
    res.status(401).json({ error: "Invalid user not logged in" });
    return;
  }
  const user = verifyToken(token, con);
  if (user === undefined) {
    res.status(401).json({ error: "Invalid user token" });
    return;
  }
  const name = req.query.name;

  if (typeof name === "string") {
    // Inserts Data into the Schedule page
    const rows = await con.all(
      "SELECT * FROM schedule WHERE userId = ?", [user]);
    // Returns a success message
    res.status(200).json(rows);
    return;
  }
  // Returns an error message with invalid parameters
  res
    .status(400)
    .json({ error: `Invalid parameters: ${JSON.stringify(req.query)}` });
}
