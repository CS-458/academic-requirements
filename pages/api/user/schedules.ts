/*
    @Description
      This is the API call that pushes the schedule form the user to the Database
*/

import { NextApiRequest, NextApiResponse } from "next";
import verifyToken from "../../../services/login";
import { userDb } from "../../../services/sql";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method !== "GET") {
    res.status(405).json({ error: "Only Get requests allowed" });
    return;
  }

  // Creates connection to the DB
  const con = await userDb();

  // Headers are lowercased
  const token = req.headers["x-google-token"];
  if (typeof token !== "string") {
    res.status(401).json({ error: "Invalid user not logged in" });
    return;
  }

  const user = await verifyToken(token, con);

  if (user === undefined) {
    res.status(401).json({ error: "Invalid user token" });
    return;
  }

  const rows = await con.all("SELECT * FROM schedule WHERE userID = ?", [user]);

  // Returns a success message
  res.status(200).json(rows);
}
