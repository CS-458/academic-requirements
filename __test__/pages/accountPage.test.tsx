import "@testing-library/jest-dom";
import { jest } from "@jest/globals";
import { screen, waitFor, within } from "@testing-library/react";
import {
  fetchApiRoute,
  mockUserInfo,
  parentEl,
  render,
  setupMockUserDB,
  setupUser
} from "../util";
import App from "../../pages/account";
import { UserLogin } from "../../services/user";
import { academicDb, userDb } from "../../services/sql";
import { savedSchedule } from "../sampleData";
import Router from "next/router";

const dbSchedule = {
  Major: 2,
  Concentration: 13,
  "Completed Courses": [],
  schedule: savedSchedule,
  usedFourYearPlan: false
};

beforeAll(async () => {
  // @ts-expect-error This is intended to override the Router.push function
  Router.push = jest.fn().mockResolvedValue(true);
  // Build mock schedule data
  await setupMockUserDB();
  const users = await userDb();
  await users.exec("INSERT INTO user VALUES ('a', 'user')");
  await users.exec(
    `INSERT INTO schedule(userID, name, scheduleData) VALUES ('a', 'Schedule 1', '${JSON.stringify(
      dbSchedule
    )}')`
  );
  await users.exec(
    `INSERT INTO schedule(userID, name, scheduleData) VALUES ('a', 'Schedule 2', '${JSON.stringify(
      dbSchedule
    )}')`
  );
  await users.exec(
    `INSERT INTO schedule(userID, name, scheduleData) VALUES ('a', 'Schedule 3', '${JSON.stringify(
      {
        Major: -1,
        Concentration: -1,
        "Completed Courses": [],
        schedule: savedSchedule,
        usedFourYearPlan: false
      }
    )}')`
  );
  await users.exec(
    `INSERT INTO schedule(userID, name, scheduleData) VALUES ('a', 'Schedule 4', '${JSON.stringify(
      {
        Major: -1,
        Concentration: 13,
        "Completed Courses": [],
        schedule: savedSchedule,
        usedFourYearPlan: false
      }
    )}')`
  );
});

test("Test Render of the account page", async () => {
  const user = setupUser();
  const bar = render(
    <UserLogin.Provider value={mockUserInfo("a")}>
      <App />
    </UserLogin.Provider>
  );
  expect(bar.baseElement).toMatchSnapshot();
  expect(screen.getByText(/Saved Schedules/i)).toBeInTheDocument();
  const schedule1 = parentEl(
    await screen.findByText(/Schedule 1/i),
    "Schedule"
  );
  expect(schedule1).toBeInTheDocument();
  await user.click(within(schedule1).getByTestId("delete"));
  await waitFor(() => expect(screen.queryByText(/Schedule 1/i)).toBeNull());
});

test("Test Edit button", async () => {
  const user = setupUser();
  const bar = render(
    <UserLogin.Provider value={mockUserInfo("a")}>
      <App />
    </UserLogin.Provider>
  );
  expect(bar.baseElement).toMatchSnapshot();
  expect(screen.getByText(/Saved Schedules/i)).toBeInTheDocument();
  const schedule = parentEl(await screen.findByText(/Schedule 2/i), "Schedule");
  expect(schedule).toBeInTheDocument();
  await user.click(within(schedule).getByTestId("edit"));
  await waitFor(() => expect(Router.push).toHaveBeenCalled());
  expect(Router.push).toHaveBeenCalledWith("/scheduler");

  const academic = await academicDb();
  expect(JSON.parse(localStorage.getItem("user_major") ?? "{}")).toMatchObject({
    completed_courses: dbSchedule["Completed Courses"],
    concentration: await academic.get(
      "select idConcentration,name,fourYearPlan  from concentration where idConcentration = ?",
      [dbSchedule.Concentration]
    ),
    major: {
      id: dbSchedule.Major,
      name: (
        await academic.get(
          "select name from major where idMajor = ?",
          dbSchedule.Major
        )
      ).name
    },
    load_four_year_plan: dbSchedule.usedFourYearPlan,
    schedule_name: "Schedule 2"
  });

  expect(
    JSON.parse(localStorage.getItem("current-schedule") ?? "{}")
  ).toMatchObject(savedSchedule);
});

test("Test Edit button w/ Invalid Schedule", async () => {
  const user = setupUser();
  render(
    <UserLogin.Provider value={mockUserInfo("a")}>
      <App />
    </UserLogin.Provider>
  );

  {
    expect(screen.getByText(/Saved Schedules/i)).toBeInTheDocument();
    const schedule1 = parentEl(
      await screen.findByText(/Schedule 3/i),
      "Schedule"
    );
    expect(schedule1).toBeInTheDocument();
    await user.click(within(schedule1).getByTestId("edit"));
    await waitFor(() => expect(screen.queryByText(/Schedule 3/i)).toBeNull());
  }
  {
    expect(screen.getByText(/Saved Schedules/i)).toBeInTheDocument();
    const schedule1 = parentEl(
      await screen.findByText(/Schedule 4/i),
      "Schedule"
    );
    expect(schedule1).toBeInTheDocument();
    await user.click(within(schedule1).getByTestId("edit"));
    await waitFor(() => expect(screen.queryByText(/Schedule 4/i)).toBeNull());
  }
});

test("Test Redirect", async () => {
  const bar = render(
    <UserLogin.Provider value={undefined}>
      <App />
    </UserLogin.Provider>
  );
  expect(bar.baseElement).toMatchSnapshot();
  await waitFor(() => expect(Router.push).toHaveBeenCalled());
  expect(Router.push).toHaveBeenCalledWith("/");
});
