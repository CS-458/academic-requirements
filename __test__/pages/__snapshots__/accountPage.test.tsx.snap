// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Edit button 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div
        class="css-m84gzu-MuiStack-root"
      >
        <div
          class="css-nen11g-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-nyqf5m-MuiTypography-root"
          >
            Saved Schedules
          </h4>
          <div
            class="css-1okj3ks-MuiStack-root"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Test Redirect 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    />
  </div>
</body>
`;

exports[`Test Render of the account page 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div
        class="css-m84gzu-MuiStack-root"
      >
        <div
          class="css-nen11g-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-nyqf5m-MuiTypography-root"
          >
            Saved Schedules
          </h4>
          <div
            class="css-1okj3ks-MuiStack-root"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
